### PR TITLE
Refine amount recalculations of traditional DEx via RPC

### LINKED ISSUE #188: Information about (traditional) DEx prices may not be accurate
To track the inline TODO and discussion of https://github.com/OmniLayer/omnicore/pull/173/files#r36941720:

When using `omni_gettransaction`, then the price/ratio returned for traditional DEx sell offers may not be accurate:

---

Alice has only 50 MSC, but offers 1000 MSC for 1000 BTC. The (aweful) way how it's currently handled: the transaction is valid, and [the offer is adjusted](https://github.com/dexX7/bitcoin/blob/d761109327539abfd44af5e050f091ae32662db8/src/omnicore/dex.cpp#L131-L147) to 50 MSC for 50 BTC. The adjusted MSC amount is [fetched from the DB](https://github.com/dexX7/bitcoin/blob/0356f8f8399b6feac7d5e0d2fe7d62ee56016972/src/omnicore/rpctxobject.cpp#L264-L265), but the BTC amount isn't updated:

```java
omni_getbalance moR6aNzFHKSMUdqKb3vDCdetM7pLBHozKY 1
{
  "balance" : "50.00000000",  // <----
  "reserved" : "0.00000000"
}
```
```java
// DEx Offer of 1000 MSC for 1000 BTC
omni_sendrawtx "moR6aNzFHKSMUdqKb3vDCdetM7pLBHozKY" "0001001400000001000000174876e800000000174876e8000a000000000098968001"
f5c71761e1f6588784720bebf1eecbe3113b99a17d4857ec776352af566293ec
```
```java
setgenerate true 1
[
  "5a52c577e14918128425d182174cd6b36eaa17d1b92b68604145c5874476cedd"
]
```
```java
omni_gettransaction f5c71761e1f6588784720bebf1eecbe3113b99a17d4857ec776352af566293ec
{
  "txid": "f5c71761e1f6588784720bebf1eecbe3113b99a17d4857ec776352af566293ec",
  "fee": "0.00010000",
  "sendingaddress": "moR6aNzFHKSMUdqKb3vDCdetM7pLBHozKY",
  "ismine": true,
  "version": 1,
  "type_int": 20,
  "type": "DEx Sell Offer",
  "propertyid": 1,
  "divisible": true,
  "amount": "50.00000000",  // <----
  "bitcoindesired": "1000.00000000",  // <---- !!
  "timelimit": 10,
  "feerequired": "0.10000000",
  "action": "new",
  "valid": true,
  "blockhash": "5a52c577e14918128425d182174cd6b36eaa17d1b92b68604145c5874476cedd",
  "blocktime": 1439467162,
  "block": 103,
  "confirmations": 1
}
```

---

As *temporary* solution (until we handle it better) I suggest to access and reuse a function to recalculate the amounts. This isn't a helper for the RPC layer, but already part of the core logic: [adjustDExOffer](https://github.com/OmniLayer/omnicore/blob/d761109327539abfd44af5e050f091ae32662db8/src/omnicore/dex.cpp#L94). We need to double-check [omni_getactivedexsells](https://github.com/OmniLayer/omnicore/blob/dd7efd7162f6dddaa04ea269c27c9dbb8b229cc4/src/omnicore/rpc.cpp#L1132-L1134) as well, which has a similar inline note.

### PR DESCRIPTION
**Pull request by dexX7**
The original DEx calculations of Master Core 0.0.9 were restored, which seems safer, if we consider all aspects and mirror it to 100 %.

As temporary solution the core calculation is exposed, so it can be used in other places. The additional "divide by zero" protection was added, because the function is now exposed.

When using `"omni_gettransaction"` in combination with a DEx sell offer, which was adjusted, because the seller offered more than available, then the amounts returned via RPC reflect the adjustment properly, and it's no longer one sided, i.e. previously only the amount for sale was adjusted, but not the desired amount. This resolves #188.

The use of `{ }` solely serves as hint, to point out that these parts shouldn't be done on the RPC layer.

More information is returned for active accept orders on the traditional DEx, when using `"omni_getactivedexsells"`. It also replaces the amount recalculations with the original function that is also used for the core logic.

### CODE REVIEW

**Comment by zathras-crypto** [comment]
Tested & confirmed passes all consensus checkpoints.  

Reviewed, confirmed calculations are as per 0.0.9.

Nice fix addressing the non-adjusted BTC amount on the RPC layer :)

I see what you mean re comments on subaction manipulation at the RPC layer.  It's awkard - manipulating subaction based on the current state (eg whether an existing sell offer exists) seemed appropriate in packet logic (ie in ```logicMath_TradeOffer```) but I tried that and of course found since packet logic is called only once during actual parsing, not later in RPC calls it doesn't address the issue.

Here https://github.com/zathras-crypto/omnicore/commit/81d0db05816f2a92dbc4e4a3f32972b1a6590791 is a small patch that moves subaction manipulation for v0 sells into ```interpret_TradeOffer```, tested and it does the job but it just seems awkward to query the state during payload decoding instead of payload logic, though I'm not sure how to get around that without doing it at the RPC layer instead...  

Re the PR, great stuff, OK to merge :)

**Comment by dexX7** [comment]
Thanks!

> Here zathras-crypto@81d0db0 is a small patch .. though I'm not sure how to get around that without doing it at the RPC layer instead

That's a neat idea! Though I also see the problem. If we think in layers, where parsing is level 0, interpretation level 1, the logic on level 2 and finally RPC on level 3, then I'd argue it would be a lesser evil, if it's done on the higher levels. Then again, on the other hand: if we had access to the subaction byte, we may simplify the [switch in the logicMath](https://github.com/zathras-crypto/omnicore/blob/81d0db05816f2a92dbc4e4a3f32972b1a6590791/src/omnicore/tx.cpp#L997-L1061) significantly. I'd further point out that we also manipulate the `CMPTransaction` in the [logicMath of simple sends](https://github.com/zathras-crypto/omnicore/blob/81d0db05816f2a92dbc4e4a3f32972b1a6590791/src/omnicore/tx.cpp#L846), which may be seen as precedence, so i'd say, ideally we'd do something like you proposed in the logicMath, however, this does not even partially solve our initial problem for unconfirmed transactions.

Ultimately, unrelated to the "how we do it", then:

- v0: we can determine whether it the `subaction` is, or is not, `cancel` at any time based on `nValue`
- v0: we can't determine whether the `subaction` is `new` or `update`, until the transaction confirms
- v1: we can determine the subaction at any time

What would you consider as ideal solution (unrelated to "how")? I tend to believe it would be to strictly differentiate between v0 and v1, because even if they are similar, they are not the same thing, so maybe the flaw is inherited, because we try to treat them the same. [Here](http://verraes.net/2014/08/dry-is-about-knowledge/) is an interesting blog post discussing "similar" things.

I'm really twisted here, hehe. I'd say, let's choose the way of the least friction for now: doing nothing, but let's go back to the topic, and take another serious look, the first time we stumble upon it, maybe in a new context or during some refactoring, where we may have new insights.

**Comment by zathras-crypto** [comment]
> I'd further point out that we also manipulate the CMPTransaction in the logicMath of simple sends, which may be seen as precedence

The difficulty here is that ```logicMath``` is not called when we retrieve a transaction on the RPC layer (it's only called during parsing).  Thus any modification of the CMPTransaction (such as changing the subaction field) would not happen during transaction retrieval via an RPC call - so we'd be back to square one in needing to manipulate at the RPC layer.

> What would you consider as ideal solution 

Well, when you actually look at it top-down, an "Update" is just a "Replace", since all we do is delete the existing sell and then list the new one.  In my ideal we would:

1) Amend the spec to state:
*From*
> For version 0 of this message and Amount for sale is non-zero, it is treated as Action=1 (New offer) unless there is already an offer outstanding from this address for the same Currency identifier, in which case it is treated as Action = 2 (Update offer).

*To*
> For version 0 of this message when Amount For Sale is non-zero, it is treated as Action=1 (New Offer).  Version 0 New Offers will automatically replace an existing sell offer.

2) Set the action byte for v0 sells to either ```1``` or ```3``` only during packet interpretation based on the amount field.

In a nutshell just remove the entire "update" concept and say v0 sells replace existing sells.  That way we remove state from the equation altogether.  I'm not sure how viable this is though, just thinking out loud...
